indev/libinput.c

```diff
@@ -259,37 +365,71 @@ void libinput_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
-      goto report_most_recent_state;
-    default:
-      break;
-  }
-  libinput_dispatch(state->libinput_context);
-  while((event = libinput_get_event(state->libinput_context)) != NULL) {
-    switch (indev_drv->type) {
-      case LV_INDEV_TYPE_POINTER:
-        read_pointer(state, event);
-        break;
-      case LV_INDEV_TYPE_KEYPAD:
-        read_keypad(state, event);
-        break;
-      default:
-        break;
-    }
-    libinput_event_destroy(event);
+  LV_UNUSED(indev_drv);
+
+  pthread_mutex_lock(&state->event_lock);
+
+  /* We may want to send this event twice so only peek it for now */
+  libinput_lv_event_t *evt = peek_event(state);
+  if (!evt)
+    evt = &state->last_event; /* indev expects us to report the most recent state */
+
+  /* Rewrite relative pointer click and motion events into absolute ones using the previous event */
+  if (evt->is_relative) {
+    /* We need to read unrotated display dimensions directly from the driver because libinput won't account
+     * for any rotation inside of LVGL */
+    lv_disp_drv_t *drv = lv_disp_get_default()->driver;
+
+    evt->point.x = state->last_event.point.x + evt->point.x;
+    evt->point.y = state->last_event.point.y + evt->point.y;
+    evt->point.x = LV_CLAMP(0, evt->point.x, drv->hor_res - 1);
+    evt->point.y = LV_CLAMP(0, evt->point.y, drv->ver_res - 1);
+    evt->is_relative = false;
+  } else if (evt->pressed == LV_INDEV_STATE_REL && evt->slot == 0
+    && state->slots[1].pressed == LV_INDEV_STATE_PR && !state->doing_mtouch_dummy_event) {
+    /*
+     * We don't support "multitouch", but libinput does. To make fast typing with two thumbs
+     * on a keyboard feel good, it's necessary to handle two fingers individually. The edge
+     * case here is if you press a key with one finger and then press a second key with another
+     * finger. No matter which finger you release, it will count as the second finger releasing
+     * and ignore the first.
+     *
+     * To work around this, detect the case where a finger is releasing while the other finger is
+     * still pressed and insert a dummy press event for the finger which is still pressed.
+     */
+
+    /* evt won't be consumed so it will be re-sent on the next call */
+    evt->pressed = LV_INDEV_STATE_PR;
+    evt->point = state->slots[0].point;
+    state->doing_mtouch_dummy_event = 1;
+
+    /* slot 1 will become slot 0 when slot 0 is released */
+    state->slots[1].pressed = LV_INDEV_STATE_REL;
+  } else if (state->doing_mtouch_dummy_event == 1) {
+    /* Now that the position is definitely correct, send the release */
+    evt->pressed = LV_INDEV_STATE_REL;
+    state->doing_mtouch_dummy_event++;
+  } else if (state->doing_mtouch_dummy_event == 2) {
+    /* Finally, update the position to the remaining finger and send a press */
+    evt->pressed = LV_INDEV_STATE_PR;
+    evt->point = state->slots[1].point; /* Wherever slot 1 most recently pressed */
+    state->doing_mtouch_dummy_event = 0;
```

Comment:
I think you can consume the event here? If I'm reading the code correctly, it'll otherwise be hit again on the next call of this function in which it should enter the `else` right below. It's probably a no-op because it's just triggering the same press event twice but probably worth avoiding.

Reply:
Nice catch! I think you're right.